### PR TITLE
WCSAxes AxisLabel refactoring

### PR DESCRIPTION
**Pull request by dstansby**
As I dive deep into the rabbit hole that is https://github.com/astropy/astropy/pull/10797, I have refactored some code along the way. This is probably easier to review one commit at a time

- Move the code to calculate the distance halfway along a spine to the spine class itself.
- Move code outside a loop that doesn't use the loop variable
- De-duplicate a check in an if statement that occurs 4 times
- Cleanup the logic used in calculating the position of tick labels

### CODE REVIEW

File: astropy/visualization/wcsaxes/axislabels.py [hunk 1]
```diff
@@ -105,37 +105,30 @@ def draw(self, renderer, bboxes, ticklabels_bbox,
                     coord_ticklabels_bbox[axis] = [None]
 
                 visible = axis in visible_ticks and coord_ticklabels_bbox[axis][0] is not None
+                x = xcen
+                y = ycen
```
Comment (Cadair) [inline astropy/visualization/wcsaxes/axislabels.py:109]: Why redfine these here? why not just call them `x` & `y` from the outset?

**Review by Cadair — APPROVED** [review]
Makes sense to me.

**Comment by pllim** [comment]
I am going to close/re-open to kick off RTD again to make sure the timeout isn't related.

**Comment by pllim** [comment]
@Cadair , if you are not waiting for someone else to review, feel free to merge when CI passes.

**Comment by Cadair** [comment]
hmmm, it timed out again.

**Comment by dstansby** [comment]
Rebased to see if the tests will pass; I also dropped the first commit, since on a second viewing it looks like that might not be a valid change to make (even though the tests passed).

**Comment by Cadair** [comment]
@dstansby should this be a bugfix as well? is it needed for the bbox quest?

**Comment by dstansby** [comment]
I think the bbox quest is kind of a bugfix, but in reality will require so much refactoring (and is a beaviour change, albeit from an incorrect to a correct behaviour) it's probably best to put it in a major release?

**Pull request merged**